File: src/simulator.js
// Synthetic document video source.
//
// Direct port of simulator.py. Renders a sequence of frames showing a
// document at various positions and orientations, so the auto-capture
// pipeline can be exercised end-to-end without a real camera.
//
// Used by:
//   - the browser app's "simulator" button (offline smoke test)
//   - the test suite for end-to-end verification

import { computeHomography, warpBilinear } from "./dewarp/homography.js";

// Real receipt pages baked from the SRD bills dataset (assets/sim/*.jpg,
// dewarped crops). Loaded async; the synthetic B&W page is the immediate
// fallback so the sim works before the fetch lands (and offline).
async function loadReceiptPage(rng) {
  const idx = await (await fetch("assets/sim/index.json")).json();
  const name = idx[Math.floor(rng() * idx.length)];
  const img = new Image();
  img.src = "assets/sim/" + name;
  await img.decode();
  const c = document.createElement("canvas");
  c.width = img.naturalWidth; c.height = img.naturalHeight;
  const ctx = c.getContext("2d");
  ctx.drawImage(img, 0, 0);
  return ctx.getImageData(0, 0, c.width, c.height);
}

/**
 * Render a clean B&W document page to an ImageData on an offscreen canvas.
 * Returns ImageData (RGBA).
 */
function drawDocumentPage(w, h, lines = 12) {
  const c = document.createElement("canvas");
  c.width = w; c.height = h;
  const ctx = c.getContext("2d");
  ctx.fillStyle = "white";
  ctx.fillRect(0, 0, w, h);

  const margin = Math.floor(Math.min(w, h) * 0.08);
  ctx.fillStyle = "black";
  ctx.fillRect(margin, margin, w - 2 * margin, Math.floor(h * 0.06));

  const lineH = Math.floor((h - 2 * margin - Math.floor(h * 0.10)) / lines);
  let y = margin + Math.floor(h * 0.12);
  ctx.fillStyle = "rgb(60,60,60)";
  for (let i = 0; i < lines; i++) {
    const lineW = Math.floor(w * (0.5 + 0.4 * (i % 3) / 2));
    ctx.fillRect(margin, y, lineW, Math.floor(lineH * 0.3));
    y += lineH;
  }
  return ctx.getImageData(0, 0, w, h);
}

/**
 * Warp the page image onto a 4-corner quad in the destination canvas.
 * Pure-JS implementation via bilinear inverse-warp. Works on every
 * browser (no DOMMatrix / setTransform(matrix) dependency, which is the
 * reason the previous version failed on iOS Chrome and older Safari).
 *
 * @param {ImageData} pageImg  source page (RGBA)
 * @param {number[]} dstQuad   flat 8-array of 4 corners
 * @param {number} outW
 * @param {number} outH
 * @returns {ImageData} the warped composite
 */
function warpPerspective(pageImg, dstQuad, outW, outH) {
  const srcH = pageImg.height, srcW = pageImg.width;

  // Use the homography module's warpBilinear — bilinear inverse-warp in
  // pure JS. iOS Chrome / Safari have intermittent bugs with creating an
  // intermediate 2D context for imageData manipulation, so we go through
  // exactly one canvas (the OUT canvas) at the end.
  const H = computeHomography(
    [0, 0, srcW - 1, 0, srcW - 1, srcH - 1, 0, srcH - 1],
    dstQuad
  );

  // Convert pageImg (RGBA Uint8ClampedArray) into RGB triples.
  // The warped RGB is built directly into our out pixels; we don't
  // stage a tmp canvas because iOS Chrome occasionally returns null
  // from getContext('2d') for tmp canvases created mid-frame.
  const srcRGB = new Uint8ClampedArray(srcW * srcH * 3);
  for (let i = 0; i < srcW * srcH; i++) {
    srcRGB[i * 3]     = pageImg.data[i * 4];
    srcRGB[i * 3 + 1] = pageImg.data[i * 4 + 1];
    srcRGB[i * 3 + 2] = pageImg.data[i * 4 + 2];
  }
  const warpedRGB = warpBilinear(srcRGB, srcH, srcW, 3, H, outH, outW);

  // Build a fresh ImageData via the OUT canvas (one stage only).
  // We pass our RGBA directly into createImageData: route one.
  const outCanvas = document.createElement("canvas");
  outCanvas.width = outW;
  outCanvas.height = outH;
  const outCtx = outCanvas.getContext("2d");
  if (!outCtx) {
    throw new Error("canvas.getContext('2d') returned null on " + outW + "x" + outH);
  }
  const imgData = outCtx.createImageData(outW, outH);
  for (let i = 0; i < outW * outH; i++) {
    imgData.data[i * 4]     = warpedRGB[i * 3];
    imgData.data[i * 4 + 1] = warpedRGB[i * 3 + 1];
    imgData.data[i * 4 + 2] = warpedRGB[i * 3 + 2];
    imgData.data[i * 4 + 3] = 255;
  }
  return imgData;
}

/**
 * SyntheticDocumentSource: emits a stream of frames simulating a
 * document in front of a camera.
 *
 * Pattern (matches the python simulator):
 *   0..30   : doc enters from the right
 *   30..90  : holds steady near the center, gentle wobble
 *   90..120 : doc moves away / out of frame
 *   120..150: blank window (cooldown)
 *   150..180: doc re-enters
 *   180..240: second hold
 *   else    : blank
 */
export class SyntheticDocumentSource {
  constructor({ width = 640, height = 480, seed = 42 } = {}) {
    this.W = width; this.H = height;
    this.frameIdx = 0;
    this.rng = mulberry32(seed);
    this.page = drawDocumentPage(Math.floor(width * 0.6), Math.floor(height * 0.75));
    this.bgCanvas = buildBackground(width, height, this.rng);
    // Swap in a real receipt from the bills dataset when it loads.
    loadReceiptPage(this.rng).then((p) => { this.page = p; }).catch(() => {});
  }

  _quadForFrame() {
    const f = this.frameIdx;
    if (120 <= f && f < 150) return null;
    let cx, cy, scale;
    if (f < 30) {
      const t = f / 30.0;
      cx = this.W * (0.95 - 0.35 * t);
      cy = this.H * 0.55;
      scale = 0.5 + 0.4 * t;
    } else if (f < 90) {
      const t = (f - 30) / 60.0;
      cx = this.W * 0.55 + 4 * Math.sin(2 * Math.PI * t * 1.5);
      cy = this.H * 0.50 + 3 * Math.cos(2 * Math.PI * t * 2.0);
      scale = 0.9;
    } else if (f < 120) {
      const t = (f - 90) / 30.0;
      cx = this.W * (0.55 - 0.4 * t);
      cy = this.H * (0.50 - 0.2 * t);
      scale = 0.9 * (1.0 - 0.2 * t);
    } else if (f < 180) {
      const t = (f - 150) / 30.0;
      cx = this.W * (0.95 - 0.40 * t);
      cy = this.H * 0.55;
      scale = 0.5 + 0.4 * t;
    } else if (f < 240) {
      const t = (f - 180) / 60.0;
      cx = this.W * 0.55 + 5 * Math.sin(2 * Math.PI * t * 1.2);
      cy = this.H * 0.50;
      scale = 0.9;
    } else {
      return null;
    }

    const skewX = 0.05 * Math.sin(f * 0.1);
    const skewY = 0.03 * Math.cos(f * 0.13);
    const docW = Math.floor(this.W * 0.55 * scale);
    const docH = Math.floor(this.H * 0.7 * scale);
    const halfW = docW / 2, halfH = docH / 2;
    return [
      cx - halfW * (1 + skewX), cy - halfH * (1 - skewY),
      cx + halfW * (1 - skewX), cy - halfH * (1 + skewY),
      cx + halfW * (1 + skewX), cy + halfH * (1 - skewY),
      cx - halfW * (1 - skewX), cy + halfH * (1 + skewY),
    ];
  }

  /**
   * Static mode for demo/testing: a single doc-in-frame image, no
   * motion. Use this so the lock-gate can satisfy its 8-stable-frame
   * debounce on a doc that isn't moving in and out of view.
   */
  staticFrame() {
    // Render once
    const frame = this.read(60);  // frame 60 = peak "hold" position
    this.frameIdx = 60;            // stay there forever
    return frame;
  }

  /**
   * Read one frame. The optional `atFrame` parameter lets the loop
   * request a specific frame (e.g. for the demo's hold phase). Otherwise
   * advances by 1.
   */
  read(atFrame) {
    if (atFrame !== undefined) this.frameIdx = atFrame;
    const out = document.createElement("canvas");
    out.width = this.W; out.height = this.H;
    const outCtx = out.getContext("2d");
    if (!outCtx) {
      // iOS Chrome occasionally returns null here when the canvas is
      // created off-screen. Surface the error so the loop can recover.
      throw new Error("simulator: canvas 2d context unavailable");
    }
    outCtx.drawImage(this.bgCanvas, 0, 0);

    const quad = this._quadForFrame();
    if (quad !== null) {
      const warped = warpPerspective(this.page, quad, this.W, this.H);
      // Composite only INSIDE the quad — warpBilinear fills the exterior
      // with black, and blitting that wholesale erased the background
      // (old sim dumps show a black bg for exactly this reason).
      const bgData = outCtx.getImageData(0, 0, this.W, this.H);
      const q = quad;
      const inQuad = (x, y) => {
        let inside = false;
        for (let i = 0, j = 3; i < 4; j = i++) {
          const xi = q[i * 2], yi = q[i * 2 + 1], xj = q[j * 2], yj = q[j * 2 + 1];
          if ((yi > y) !== (yj > y) && x < (xj - xi) * (y - yi) / (yj - yi) + xi) inside = !inside;
        }
        return inside;
      };
      for (let y = 0; y < this.H; y++) {
        for (let x = 0; x < this.W; x++) {
          if (!inQuad(x, y)) continue;
          const i = (y * this.W + x) * 4;
          bgData.data[i] = warped.data[i];
          bgData.data[i + 1] = warped.data[i + 1];
          bgData.data[i + 2] = warped.data[i + 2];
        }
      }
      outCtx.putImageData(bgData, 0, 0);
    }
    this.frameIdx += 1;
    return { ok: true, imageData: outCtx.getImageData(0, 0, this.W, this.H) };
  }

  release() {
    this.bgCanvas = null;
  }
}

function buildBackground(W, H, rng) {
  // Wood-grain-ish table: warm base, horizontal streaks (low-freq per-row
  // luma), plank seams, per-pixel noise. Matches the real-corpus failure
  // modes far better than the old flat dark fill.
  const c = document.createElement("canvas");
  c.width = W; c.height = H;
  const ctx = c.getContext("2d");
  const img = ctx.createImageData(W, H);
  const rowTone = new Float32Array(H);
  let t = 0;
  for (let y = 0; y < H; y++) {
    t = 0.92 * t + (rng() - 0.5) * 6;                    // wandering grain
    rowTone[y] = 18 * Math.sin(y * 0.05) + t;
  }
  const seam = Math.floor(H * (0.3 + 0.4 * rng()));      // one plank seam
  for (let y = 0; y < H; y++) {
    const grain = rowTone[y] + (Math.abs(y - seam) < 2 ? -35 : 0);
    for (let x = 0; x < W; x++) {
      const n = (rng() - 0.5) * 14;
      const streak = 10 * Math.sin(x * 0.01 + y * 0.2);
      const i = (y * W + x) * 4;
      img.data[i]     = clamp(120 + grain + streak + n);   // R
      img.data[i + 1] = clamp(78 + grain * 0.8 + streak * 0.7 + n);  // G
      img.data[i + 2] = clamp(48 + grain * 0.6 + streak * 0.5 + n);  // B
      img.data[i + 3] = 255;
    }
  }
  ctx.putImageData(img, 0, 0);
  return c;
}

function clamp(v) {
  return v < 0 ? 0 : v > 255 ? 255 : v;
}

// Deterministic RNG matching numpy's default_rng behavior for the
// small uses here (just integer draws via floor(rng()*N)). mulberry32
// is a 32-bit PRNG that gives stable sequences across runs.
function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a = (a + 0x6D2B79F5) | 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

File: simulator.py
"""Synthetic document video source.

Generates a sequence of frames showing a document at various positions
and orientations, so the auto-capture pipeline can be exercised end to
end without a real camera.

The synthetic document is rendered onto a contrasting background. It
``approaches`` the camera, holds steady, then moves out. The pipeline
should fire during the steady hold.

This is a self-test for the detector + tracker + dewarp chain.
"""

from __future__ import annotations
import numpy as np
from PIL import Image, ImageDraw, ImageFont


def _draw_document_page(w: int, h: int, lines: int = 12) -> np.ndarray:
    """Render a clean B&W-ish document page on a white background."""
    img = Image.new("RGB", (w, h), "white")
    draw = ImageDraw.Draw(img)
    # margin
    margin = int(min(w, h) * 0.08)
    # title bar
    draw.rectangle([margin, margin, w - margin, margin + int(h * 0.06)], fill="black")
    # body lines
    line_h = int((h - margin * 2 - int(h * 0.10)) / lines)
    y = margin + int(h * 0.12)
    for i in range(lines):
        line_w = int(w * (0.5 + 0.4 * (i % 3) / 2))
        draw.rectangle([margin, y, margin + line_w, y + int(line_h * 0.3)],
                       fill=(60, 60, 60))
        y += line_h
    arr = np.asarray(img)
    return arr[..., ::-1].copy()  # RGB -> BGR


def _warp_perspective(src: np.ndarray, dst_quad: np.ndarray,
                      out_size: tuple[int, int]) -> np.ndarray:
    """Pure-numpy perspective warp (delegates to the project module)."""
    from autocapture.dewarp.homography import compute_homography, warp_bilinear
    h, w = out_size
    src_h, src_w = src.shape[:2]
    # The src quad is the page corners in src order
    src_quad = np.array(
        [[0, 0], [src_w - 1, 0], [src_w - 1, src_h - 1], [0, src_h - 1]],
        dtype=np.float64
    )
    H = compute_homography(src_quad, dst_quad.astype(np.float64))
    return warp_bilinear(src, H, h, w)


class SyntheticDocumentSource:
    """Emits a stream of frames simulating a document in front of a camera.

    The script is deterministic given a seed. Pattern:
      - frames 0..30:    doc enters from the right, off-center
      - frames 30..90:   doc holds steady near the center, slight wobble
      - frames 90..120:  doc moves away / out of frame
      - frames 120..150: blank frame (cooldown)
      - frames 150..210: doc re-enters and holds again
    """

    def __init__(self, width: int = 640, height: int = 480, seed: int = 42):
        self.W, self.H = width, height
        self.frame_idx = 0
        self.rng = np.random.default_rng(seed)
        self._page = _draw_document_page(int(width * 0.6), int(height * 0.75))
        self._bg_color = (40, 30, 25)  # dark wood-ish, BGR

    def _background(self) -> np.ndarray:
        bg = np.full((self.H, self.W, 3), self._bg_color, dtype=np.uint8)
        # Add a tiny bit of texture (deterministic)
        noise = self.rng.integers(-10, 10, size=bg.shape, dtype=np.int16)
        bg = np.clip(bg.astype(np.int16) + noise, 0, 255).astype(np.uint8)
        return bg

    def _quad_for_frame(self) -> np.ndarray | None:
        """Return the 4 corners of the doc in this frame, or None if out."""
        f = self.frame_idx
        # blank window
        if 120 <= f < 150:
            return None
        # entry phase 0..30
        if f < 30:
            t = f / 30.0
            cx = self.W * (0.95 - 0.35 * t)
            cy = self.H * 0.55
            scale = 0.5 + 0.4 * t
        # hold phase 30..90
        elif f < 90:
            t = (f - 30) / 60.0
            # gentle wobble
            cx = self.W * 0.55 + 4 * np.sin(2 * np.pi * t * 1.5)
            cy = self.H * 0.50 + 3 * np.cos(2 * np.pi * t * 2.0)
            scale = 0.9
        # exit phase 90..120
        elif f < 120:
            t = (f - 90) / 30.0
            cx = self.W * (0.55 - 0.4 * t)
            cy = self.H * (0.50 - 0.2 * t)
            scale = 0.9 * (1.0 - 0.2 * t)
        # second entry 150..180
        elif f < 180:
            t = (f - 150) / 30.0
            cx = self.W * (0.95 - 0.40 * t)
            cy = self.H * 0.55
            scale = 0.5 + 0.4 * t
        # second hold 180..240
        elif f < 240:
            t = (f - 180) / 60.0
            cx = self.W * 0.55 + 5 * np.sin(2 * np.pi * t * 1.2)
            cy = self.H * 0.50
            scale = 0.9
        else:
            return None

        # small perspective skew
        skew_x = 0.05 * np.sin(f * 0.1)
        skew_y = 0.03 * np.cos(f * 0.13)
        doc_w = int(self.W * 0.55 * scale)
        doc_h = int(self.H * 0.7 * scale)
        half_w, half_h = doc_w / 2, doc_h / 2
        quad = np.array([
            [cx - half_w * (1 + skew_x), cy - half_h * (1 - skew_y)],
            [cx + half_w * (1 - skew_x), cy - half_h * (1 + skew_y)],
            [cx + half_w * (1 + skew_x), cy + half_h * (1 - skew_y)],
            [cx - half_w * (1 - skew_x), cy + half_h * (1 + skew_y)],
        ], dtype=np.float64)
        return quad

    def read(self) -> tuple[bool, np.ndarray]:
        frame = self._background()
        quad = self._quad_for_frame()
        if quad is not None:
            doc_h, doc_w = self._page.shape[:2]
            # Use the quad to warp the page onto the background
            # The src quad in the page is the full page rect.
            warped = _warp_perspective(self._page, quad, (self.H, self.W))
            # alpha-composite: where the warped page is brighter than the bg
            # Note: paper is white (luma 255) so the upper bound must be
            # inclusive — the v1 < 250 mask was hiding the paper body.
            mask = (warped.mean(axis=2) > 30) & (warped.mean(axis=2) <= 255)
            mask3 = np.stack([mask, mask, mask], axis=-1)
            frame = np.where(mask3, warped, frame)
        self.frame_idx += 1
        return True, frame

    def release(self) -> None:
        pass
